File: golang/graph_bfs/min_generic_mut.go
package main

func minMutation(start string, end string, bank []string) int {
	bankSet := make(map[string]bool)

	for _, gene := range bank {
		bankSet[gene] = true
	}

	if !bankSet[end] {
		return -1
	}

	genes := []byte{'A', 'C', 'G', 'T'}
	queue := []struct {
		gene  string
		steps int
	}{{start, 0}}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		if current.gene == end {
			return current.steps
		}

		for i := 0; i < len(current.gene); i++ {
			for _, g := range genes {

				mutatedGene := []byte(current.gene)
				mutatedGene[i] = g
				mutated := string(mutatedGene)

				if bankSet[mutated] {
					queue = append(queue, struct {
						gene  string
						steps int
					}{mutated, current.steps + 1})
					delete(bankSet, mutated)
				}
			}
		}
	}

	return -1
}

File: java/src/main/java/com/sanedge/interview/graph_bfs/MinGenericMut.java
package com.sanedge.interview.graph_bfs;

import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

public class MinGenericMut {

    private static class GeneMutation {

        String gene;
        int steps;

        GeneMutation(String gene, int steps) {
            this.gene = gene;
            this.steps = steps;
        }
    }

    public int minMutation(String start, String end, String[] bank) {
        Set<String> bankSet = new HashSet<>(Arrays.asList(bank));

        if (!bankSet.contains(end)) {
            return -1;
        }

        char[] genes = { 'A', 'C', 'G', 'T' };

        Queue<GeneMutation> queue = new LinkedList<>();
        queue.offer(new GeneMutation(start, 0));

        while (!queue.isEmpty()) {
            GeneMutation current = queue.poll();

            if (current.gene.equals(end)) {
                return current.steps;
            }

            for (int i = 0; i < current.gene.length(); i++) {
                for (char g : genes) {
                    char[] mutatedGene = current.gene.toCharArray();

                    mutatedGene[i] = g;

                    String mutated = new String(mutatedGene);

                    if (bankSet.contains(mutated)) {
                        queue.offer(
                            new GeneMutation(mutated, current.steps + 1)
                        );
                        bankSet.remove(mutated);
                    }
                }
            }
        }

        return -1;
    }
}
